src/runtime/test_case_runner.js

```diff
@@ -166,26 +172,40 @@ export default class TestCaseRunner {
 
   async run() {
     this.emitPrepared()
-    this.emit('test-case-started', {})
-    await this.runHooks(
-      this.beforeHookDefinitions,
-      {
-        sourceLocation: this.testCaseSourceLocation,
-        pickle: this.testCase.pickle,
-      },
-      true
-    )
-    await this.runSteps()
-    await this.runHooks(
-      this.afterHookDefinitions,
-      {
-        sourceLocation: this.testCaseSourceLocation,
-        pickle: this.testCase.pickle,
-        result: this.result,
-      },
-      false
-    )
-    this.emit('test-case-finished', { result: this.result })
+    let retries = 0
+    for (; retries <= this.retries; retries++) {
+      this.emit('test-case-started', {})
+      if (retries > 0) {
+        this.resetTestProgressData()
+        this.result.retryAttempt = retries
+      }
+      await this.runHooks(
+        this.beforeHookDefinitions,
+        {
+          sourceLocation: this.testCaseSourceLocation,
+          pickle: this.testCase.pickle,
+        },
+        true
+      )
+      await this.runSteps()
+      await this.runHooks(
+        this.afterHookDefinitions,
+        {
+          sourceLocation: this.testCaseSourceLocation,
+          pickle: this.testCase.pickle,
+          result: this.result,
+        },
+        false
+      )
+      const shouldRetryTestCase =
+        this.result.status === Status.FAILED && retries < this.retries
+      if (!shouldRetryTestCase) {
+        this.emit('test-case-finished', { result: this.result })
+        break
+      }
+      this.result.status = Status.FLAKY
```

Comment:
I'm still questioning this status as its not flaky unless a later attempt passes. Thoughts on instead adding an additional boolean to the result `retried` instead of introducing this status?

Reply:
You dont know if you have retried until you retry. But we cannot mark it as failed because it will fake the whole report...
Can we mark it as "retried" only if "retries" and "number of retries" are in use?

Reply:
I've renamed `FLAKY` status into `RETRY` to be semantically more correct.